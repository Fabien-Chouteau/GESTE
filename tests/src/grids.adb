with GESTE;
with Ada.Text_IO;
with Console_Char_Screen;

procedure Grids is

   package Console_Screen is new Console_Char_Screen
     (Width       => 20,
      Height      => 20,
      Buffer_Size => 256,
      Init_Char   => ' ');

   Palette : aliased constant GESTE.Palette_Type :=
     ('#', '0', 'T', ' ');

   Background : constant Character := '_';

   Tiles : aliased constant GESTE.Tile_Array :=
     (1 => ((1, 1, 1, 1, 1),
            (1, 3, 3, 3, 1),
            (1, 3, 3, 3, 1),
            (1, 3, 3, 3, 1),
            (1, 1, 1, 1, 1)),

      2 => ((2, 3, 3, 3, 2),
            (3, 2, 3, 2, 3),
            (3, 3, 2, 3, 3),
            (3, 2, 3, 2, 3),
            (2, 3, 3, 3, 2)),

      3 => ((3, 3, 1, 3, 3),
            (3, 3, 1, 3, 3),
            (1, 1, 2, 1, 1),
            (3, 3, 1, 3, 3),
            (3, 3, 1, 3, 3))
     );

   Bank : aliased GESTE.Tile_Bank.Instance (Tiles'Unrestricted_Access,
                                                    GESTE.No_Collisions,
                                                    Palette'Unrestricted_Access);

   Grid_Data : aliased constant GESTE.Grid_Data :=
     ((0, 2, 0),
      (3, 1, 3),
      (0, 2, 0));

   Grid : aliased GESTE.Grid.Instance (Grid_Data'Unrestricted_Access,
                                               Bank'Unrestricted_Access);

begin

   Console_Screen.Verbose;

   Grid.Move ((1, 1));
   GESTE.Add (Grid'Unrestricted_Access, 0);

   GESTE.Render_Window
     (Window           => Console_Screen.Screen_Rect,
      Background       => Background,
      Buffer           => Console_Screen.Buffer,
      Push_Pixels      => Console_Screen.Push_Pixels'Unrestricted_Access,
      Set_Drawing_Area => Console_Screen.Set_Drawing_Area'Unrestricted_Access);

   Console_Screen.Print;

   Grid.Move ((5, 5));

   GESTE.Render_Dirty
     (Screen_Rect      => Console_Screen.Screen_Rect,
      Background       => Background,
      Buffer           => Console_Screen.Buffer,
      Push_Pixels      => Console_Screen.Push_Pixels'Unrestricted_Access,
      Set_Drawing_Area => Console_Screen.Set_Drawing_Area'Unrestricted_Access);

   Console_Screen.Print;

end Grids;
