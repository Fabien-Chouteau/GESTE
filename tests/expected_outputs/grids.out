New drawing area:
X Start: 0
X End: 19
Y Start: 0
Y end: 19
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ # # 0 # # _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ 0 _ _ _ 0 # # # # # 0 _ _ _ 0 _ _ _ _
_ _ 0 _ 0 _ # _ _ _ # _ 0 _ 0 _ _ _ _ _
_ _ _ 0 _ _ # _ _ _ # _ _ 0 _ _ _ _ _ _
_ _ 0 _ 0 _ # _ _ _ # _ 0 _ 0 _ _ _ _ _
_ 0 _ _ _ 0 # # # # # 0 _ _ _ 0 _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ # # 0 # # _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ # _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
New drawing area:
X Start: 1
X End: 19
Y Start: 1
Y end: 19
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ # # 0 # # _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ 0 _ _ _ 0 # # # # # 0 _ _ _ 0
_ _ _ _ _ _ 0 _ 0 _ # _ _ _ # _ 0 _ 0 _
_ _ _ _ _ _ _ 0 _ _ # _ _ _ # _ _ 0 _ _
_ _ _ _ _ _ 0 _ 0 _ # _ _ _ # _ 0 _ 0 _
_ _ _ _ _ 0 _ _ _ 0 # # # # # 0 _ _ _ 0
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ # # 0 # # _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ # _ _ _ _ _ _ _
